crates/compiler/gen_dev/src/generic64/aarch64.rs

```diff
@@ -2302,6 +2708,320 @@ fn udiv_reg64_reg64_reg64(
+#[inline(always)]
+fn fcmp_freg_freg(
+    buf: &mut Vec<'_, u8>,
+    ftype: FloatWidth,
+    src1: AArch64FloatReg,
+    src2: AArch64FloatReg,
+) {
+    let inst = FloatingPointCompare::new(FloatingPointCompareParams {
+        ptype: ftype,
+        rn: src1,
+        rm: src2,
+        opcode2: 0b00000,
+    });
+
+    buf.extend(inst.bytes());
+}
+
+/// `FCVT Sd, Dn` -> Convert 64-bit float Dn to 32-bit float Sd.
+#[inline(always)]
+fn fcvt_freg32_freg64(buf: &mut Vec<'_, u8>, dst: AArch64FloatReg, src: AArch64FloatReg) {
+    let inst =
+        FloatingPointDataProcessingOneSource::new(FloatingPointDataProcessingOneSourceParams {
+            opcode: 0b000100,
+            ptype: FloatWidth::F64,
+            rd: dst,
+            rn: src,
+        });
+
+    buf.extend(inst.bytes());
+}
+
+/// `FCVT Dd, Sn` -> Convert 32-bit float Sn to 64-bit float Dd.
+#[inline(always)]
+fn fcvt_freg64_freg32(buf: &mut Vec<'_, u8>, dst: AArch64FloatReg, src: AArch64FloatReg) {
+    let inst =
+        FloatingPointDataProcessingOneSource::new(FloatingPointDataProcessingOneSourceParams {
+            opcode: 0b000101,
+            ptype: FloatWidth::F32,
+            rd: dst,
+            rn: src,
+        });
+
+    buf.extend(inst.bytes());
+}
+
+/// `FDIV Sd/Dd, Sn/Dn, Sm/Dm` -> Divide Sn/Dn by Sm/Dm and place the result into Sd/Dd.
+#[inline(always)]
+fn fdiv_freg_freg_freg(
+    buf: &mut Vec<'_, u8>,
+    ftype: FloatWidth,
+    dst: AArch64FloatReg,
+    src1: AArch64FloatReg,
+    src2: AArch64FloatReg,
+) {
+    let inst =
+        FloatingPointDataProcessingTwoSource::new(FloatingPointDataProcessingTwoSourceParams {
+            opcode: 0b0001,
+            ptype: ftype,
+            rd: dst,
+            rn: src1,
+            rm: src2,
+        });
+
+    buf.extend(inst.bytes());
+}
+
+/// `FMOV Sd/Dd, Sn/Dn` -> Move Sn/Dn to Sd/Dd.
+#[inline(always)]
+fn fmov_freg_freg(
+    buf: &mut Vec<'_, u8>,
+    ftype: FloatWidth,
+    dst: AArch64FloatReg,
+    src: AArch64FloatReg,
+) {
+    let inst =
+        FloatingPointDataProcessingOneSource::new(FloatingPointDataProcessingOneSourceParams {
+            opcode: 0b000000,
+            ptype: ftype,
+            rd: dst,
+            rn: src,
+        });
+
+    buf.extend(inst.bytes());
+}
+
+/// Encode a 32-bit float into an 8-bit immediate for FMOV.
+/// See Table C2-1 in the ARM manual for a table of every float that can be encoded in 8 bits.
+/// If the float cannot be encoded, return None.
+/// This operation is the inverse of VFPExpandImm in the ARM manual.
+#[inline(always)]
+fn encode_f32_to_imm8(imm: f32) -> Option<u8> {
+    let n = 32;
+    let e = 8; // number of exponent bits in a 32-bit float
+    let f = n - e - 1; // 23: number of fraction bits in a 32-bit float
+
+    let bits = imm.to_bits();
+
+    let sign = (bits >> (n - 1)) & 1; // bits<31>
+    let exp = (bits >> f) & ((1 << e) - 1); // bits<30:23>
```

Comment:
Optional, but I think it may be more readable to replace these types of expressions `((1 << e) - 1)` with constants like `0xF`. Not sure if this will pan out in practice due to more complex expressions like `((1 << (e - 3)) - 1)`

Reply:
Here's what it looks like in `encode_f64_to_imm8` with that change:

```rust
    let exp = (bits >> f) & 0x7FF; // bits<62:52>
    let frac = bits & 0xFFFFFFFFFFFFF; // bits<51:0>

    let exp_first = (exp >> (e - 1)) & 1; // exp<10>
    let exp_middle = (exp >> 2) & 0xFF; // exp<9:2>
    let exp_last = exp & 0b11; // exp<0:1>
```

I don't really have a great sense of which is more readable. The way I'm currently doing it has the advantage that it's exactly the same in both `encode` functions, since it uses the `n`, `e`, and `f` constants (which I took from `VFPExpandImm` in the ARM manual). But using the hex constants might be simpler. I'm fine with whichever one you think.

Reply:
Let's just leave it as is.